### PR TITLE
Change to ntbk A and B

### LINKED ISSUE #4: Napari widget should work for nD images
Currently only works for 2d time-lapse, but we should be able to export 3d time-lapse, multichannel data.  Need to update `get_image_patches` to enable this.

### PR DESCRIPTION
**Pull request by KristinaUlicna**
- Add protocol abstract to notebooks/README.md
- Change ntbk A `get_image_patch` function so it can deal with nD data
- Modify ntbk B so it uses cellx loader & `read_annotations` function
- Resolves #4

### CODE REVIEW

**Comment by quantumjot** [comment]
@KristinaUlicna - can you resolve these conflicts first, then I'll review.

File: notebooks/README.md [hunk 1]
```diff
@@ -1,6 +1,11 @@
 # Notebooks & Repository Structure
 
-![Protocol Pipeline](../assets/repository_structure.png)
+## Abstract
```
Comment (quantumjot) [inline notebooks/README.md:3]: nice!

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 2]
```diff
@@ -204,32 +205,34 @@
...
-    "        y_coo += pad_width\n",
-    "        x_coo += pad_width\n",
-    "        padded = \"_flagged\"\n",
-    "    \n",
-    "    # Crop patch around the point:\n",
-    "    patch = np.array([row[x_coo - pad : x_coo + pad] for row in image[y_coo - pad : y_coo + pad]]).astype(np.uint8)\n",
+    "    patches = [[], False]\n",
+    "    square = shape // 2\n",
+    "\n",
+    "    for channel in range(len(layers)):\n",
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]: Why not iterate over the layers?
```suggestion
    "    for layer in layers:\n",
```

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 3]
```diff
@@ -204,32 +205,34 @@
...
+    "        \n",
+    "        # Read the image from layers:\n",
+    "        frame, y_coo, x_coo = [int(coo) for coo in coords]\n",
+    "        image = layers[channel].data[frame]\n",
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]: Then this becomes:
```suggestion
    "        image = layer.data[frame]\n",
```

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 4]
```diff
@@ -204,32 +205,34 @@
...
     "    \n",
-    "    # Check if patches are of specified shape:\n",
-    "    if patch.shape != (shape, shape):\n",
-    "        raise ValueError(f\"Image patch doesn't have correct shape: {patch.shape}\")\n",
+    "    # Check if all patches are of specified shape:\n",
+    "    if not all([patch.shape == (shape, shape) for patch in patches[0]]):\n",
+    "        raise ValueError(f\"Image patches don't have correct shape: {[patch.shape for patch in patches[0]]}\")\n",
     "    \n",
-    "    return [patch, padded]\n"
+    "    return patches\n"
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]: Here you could return:
```suggestion
    "    return np.stack(patches, axis=-1), flagged\n"
```

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 5]
```diff
@@ -204,32 +205,34 @@
...
-    "    if any([item < pad for item in edge_dist]):\n",
-    "        pad_width = int(pad - min(edge_dist))\n",
-    "        image = np.pad(array=image, pad_width=pad_width, mode='constant')\n",
-    "        y_coo += pad_width\n",
-    "        x_coo += pad_width\n",
-    "        padded = \"_flagged\"\n",
-    "    \n",
-    "    # Crop patch around the point:\n",
-    "    patch = np.array([row[x_coo - pad : x_coo + pad] for row in image[y_coo - pad : y_coo + pad]]).astype(np.uint8)\n",
+    "    patches = [[], False]\n",
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]: I don't really like this patches list, why not have two variables `patches` and `flagged` so you can return the tuple

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 6]
```diff
@@ -304,12 +310,19 @@
...
+    "                if image_patches[1] == True:\n",
+    "                    flagged += \"_flagged\"\n",
+    "                    export_data['flagged'][idx] = \"True\"\n",
+    "                \n",
     "                image_patch_fn = f\"{patch_label}/{patch_label}_{SESSION_TIME}_{idx}{flagged}.tif\"\n",
     "                \n",
     "                # open a stream to write to the zip file\n",
     "                stream = io.BytesIO()\n",
-    "                imsave(stream, image_patch, format='tif')\n",
+    "                imsave(stream, image_patches[0][0], format='tif') # TODO: Currently work on 2D time-lapse data only! \n",
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]: If you make the changes above, then this line doesn't need to change.

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 7]
```diff
@@ -304,12 +310,19 @@
...
     "\n",
     "                # grab the image patch\n",
-    "                image_patch, flagged = get_image_patch(image_layers, patch_coords, shape=shape)\n",
+    "                image_patches = get_image_patch(image_layers, patch_coords, shape=shape)\n",
+    "                \n",
+    "                # check if the item was flagged:\n",
+    "                flagged = \"\"\n",
+    "                if image_patches[1] == True:\n",
+    "                    flagged += \"_flagged\"\n",
+    "                    export_data['flagged'][idx] = \"True\"\n",
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]: Can you not just use the boolean flag here? Or does that not work with the json writer?

**Review by quantumjot — CHANGES_REQUESTED** [review]
Looks great @KristinaUlicna - I've only looked at notebook A and made some suggestions for generalisation. Could @laureho, @nthndy or @chris-soelistyo review the new notebook B?

File: notebooks/A_CNN_Annotation_with_Napari.ipynb (hunk 4 above)
Comment (quantumjot) [reply notebooks/A_CNN_Annotation_with_Napari.ipynb]: This is in reference to this comment: https://github.com/lowe-lab-ucl/cnn-annotator/issues/4#issuecomment-799372152

**Comment by nthndy** [comment]
> Looks great @KristinaUlicna - I've only looked at notebook A and made some suggestions for generalisation. Could @laureho, @nthndy or @chris-soelistyo review the new notebook B?

I successfully ran notebook B on the supplied annotation data and my own annotation data.

**Review by laureho — COMMENTED** [review]
(The diff was too big, so I couldn't make in-line suggestions. Sorry if review comments end up looking a bit messy.)

In the new notebook B `tqdm` isn't used anymore. Here I think it makes most sense to add it when plotting the image patches (last cell), like `for i in tqdm(range(len(list(states.keys())))):`

After plotting the patches with `tqdm`, I noticed that this cell is fairly resource-intensive, consistently taking almost 1 min to finish (as seen in below screenshot). And this was with an example annotation file containing only 76 images, so even when plotting a majority of random noise patches the process takes quite a while.
image
Maybe it would be sufficient to only show 3 rows of 20 example patches for a quicker visualization? In my trials, this keeps the computation time under 30 sec, which I think is more reasonable.

**Comment by KristinaUlicna** [comment]
Great suggestion - I forgot to time the cell execution with `tqdm`! Adding it now. 

I just submitted a PR to `cellx` which allows the user to check an individual zip file, too, so the plotting is not too long. Check this: https://github.com/quantumjot/cellx/issues/29#issue-831738002 

Will add the `tqdm` to inform the user about the time and other changes as soon as the PR is approved.

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 8]
```diff
@@ -197,39 +206,42 @@
...
     "    \n",
-    "    # Check if patches are of specified shape:\n",
-    "    if patch.shape != (shape, shape):\n",
-    "        raise ValueError(f\"Image patch doesn't have correct shape: {patch.shape}\")\n",
+    "    # Check if all patches are of specified shape:\n",
+    "    #if not all([patch.shape == (shape, shape) for patch in patches[0]]):\n",
+    "    #    raise ValueError(f\"Image patches don't have correct shape: {[patch.shape for patch in patches[0]]}\")\n",
     "    \n",
-    "    return [patch, padded]\n"
+    "    return patches, flagged\n",
```
Comment (KristinaUlicna) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]: I've edited the code according the instructions now. Seems to work pretty well.

Comment (laureho) [reply notebooks/A_CNN_Annotation_with_Napari.ipynb]: Just a detail, if the output isn't useful for the user, might be good to go to the "Kernel" tab and choose "Restart & Clear Output" to reset the notebook to a clean state.

**Review by laureho — APPROVED** [review]
Also a couple minor typos:
- "Read the annotated images and their respective labels & calculate the proportion of _**un/flagges**_ patches per each class label."
- "Choose whether you'd like to save out the figures to be visualised below. **_There_** will be saved in the default directory: ../assets/"

Notebook B looks great otherwise!

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 9]
```diff
@@ -197,39 +206,42 @@
...
-    "        image = np.pad(array=image, pad_width=pad_width, mode='constant')\n",
-    "        y_coo += pad_width\n",
-    "        x_coo += pad_width\n",
-    "        padded = \"_flagged\"\n",
-    "    \n",
-    "    # Crop patch around the point:\n",
-    "    patch = np.array([row[x_coo - pad : x_coo + pad] for row in image[y_coo - pad : y_coo + pad]]).astype(np.uint8)\n",
+    "    flagged = False\n",
+    "    square = shape // 2\n",
+    "\n",
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb:218]: You need to store the crop from each layer. At the moment, they're being overwritten and you're only passing the patch for the last layer in return
```suggestion
    "    patches = []\n",
```

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 10]
```diff
@@ -197,39 +206,42 @@
...
+    "        # Check if padding is needed:\n",
+    "        if any([item < square for item in edge_dist]):\n",
+    "            pad_width = int(square - min(edge_dist))\n",
+    "            image = np.pad(array=image, pad_width=pad_width, mode='constant')\n",
+    "            y_coo = y_coo + pad_width\n",
+    "            x_coo = x_coo + pad_width\n",
+    "            flagged = True\n",
+    "\n",
+    "        # Slice the patch from the image:\n",
+    "        patches = image[y_coo-square : y_coo+square, x_coo-square : x_coo+square].astype(np.uint8)\n",
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]:
```suggestion
    "        patch = image[y_coo-square : y_coo+square, x_coo-square : x_coo+square].astype(np.uint8)\n",
```

File: notebooks/A_CNN_Annotation_with_Napari.ipynb [hunk 11]
```diff
@@ -197,39 +206,42 @@
...
+    "        patches = np.stack(patches, axis=-1)\n",
```
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]:
```suggestion
    "        patches.append(patch)\n",
```

File: notebooks/A_CNN_Annotation_with_Napari.ipynb (hunk 8 above)
Comment (quantumjot) [inline notebooks/A_CNN_Annotation_with_Napari.ipynb]:
```suggestion
    "    return np.stack(patches, axis=-1), flagged\n",
```

**Review by quantumjot — CHANGES_REQUESTED** [review]
At the moment, your code won't work properly. You need to store the patches for each layer.

**Review by quantumjot — APPROVED** [review]
I'll approve this and merge. We can fix any bugs with a follow up PR.

**Pull request merged**